### PR TITLE
Update Mock classes with new error handling

### LINKED ISSUE #31520: Testerina mocking errors are not properly shown due to a typecast issue in constructed error
**Description:**
When there are invalid scenarios, the error rendering throws a typecast error.

Eg:
```
 [fail] testDefaultTooManyArgs:

      {ballerina}TypeCastError {"message":"incompatible types: 'error' cannot be cast to 'ballerina/test:0.8.0:Error?'"}

           at ballerina.test.0_8_0:validateArgumentsExt(mock.bal:341)
           ballerina.test.0_8_0.MemberFunctionStub:withArguments(mock.bal:127)
           intg_tests.object_mocking.0_0_0.tests.main_error_test:testDefaultTooManyArgs(tests/main_error_test.bal:49)

```
**Steps to reproduce:**

Run MockTests [1] in ballerina-lang manually

[1] https://github.com/ballerina-platform/ballerina-lang/tree/master/tests/testerina-integration-test/src/test/resources/project-based-tests/object-mocking-tests

**Affected Versions:**
SL Beta.1

### PR DESCRIPTION
**Pull request by Ibaqu**
Fixes #31520

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/FunctionMock.java [hunk 1]
```diff
@@ -186,38 +196,48 @@ private static void validateFunctionSignature(Method mockMethod, Method original
...
         } else {
-            throw ErrorCreator.createDistinctError(MockConstants.FUNCTION_NOT_FOUND_ERROR,
-                                                   MockConstants.TEST_PACKAGE_ID,
-                                                   StringUtils.fromString(
-                                                            "Mock function \'" + mockMethodName + "\' " +
-                                                                    "cannot be found"));
+            throw ErrorCreator.createError(
+                    MockConstants.TEST_PACKAGE_ID,
+                    MockConstants.FUNCTION_SIGNATURE_MISMATCH_ERROR,
+                    StringUtils.fromString("Mock function \'" + mockMethodName + "\' " +
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/FunctionMock.java:237]:
```suggestion
                    StringUtils.fromString("Mock function '" + mockMethodName + "' " +
```
Why do we need to escape single quotes?

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/ObjectMock.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
  */
```
Comment (pramodya1994) [inline misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/ObjectMock.java:17]: There are some unused class definitions in this class.
Shall we remove them if we don't use them anywhere?
eg: `validateFunctionName` in L118, `validateFieldName` in L139

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java [hunk 3]
```diff
@@ -61,11 +61,93 @@ public void testFunctionMocking() throws BallerinaTestException {
     public void testObjectMocking() throws BallerinaTestException {
         String msg1 = "7 passing";
         String msg2 = "6 failing";
+
```
Comment (pramodya1994) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java:64]:
```suggestion
```
Do we need a new line inbetween?

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java [hunk 4]
```diff
@@ -61,11 +61,93 @@ public void testFunctionMocking() throws BallerinaTestException {
...
+        String msg = "too many argument provided to mock the function get()";
+        String[] args = mergeCoverageArgs(new String[]{"--tests", "object_mocking:testDefaultTooManyArgs"});
+        String output =
+                balClient.runMainAndReadStdOut("test", args, new HashMap<>(),
+                        projectBasedTestsPath.resolve("object-mocking-tests").toString(), false);
+        if (!output.contains(msg)) {
+            throw new BallerinaTestException(
+                    "Test failed due to default module single test failure.\nOutput:\n" + output);
+        }
+    }
```
Comment (pramodya1994) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java:152]: Shall we write a single parameterized test using data providers for this test if possible?
Check `Passing Parameters with Dataproviders` part in [this tutorial](https://www.tutorialspoint.com/testng/testng_parameterized_test.htm).

**Review by pramodya1994 — CHANGES_REQUESTED** [review]
Added few suggetions

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/ObjectMock.java (hunk 2 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/ObjectMock.java:17]: `validateFunctionName` and `validateFieldName` are used as part of the interop in `mock.bal`

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java [hunk 5]
```diff
@@ -61,11 +61,93 @@ public void testFunctionMocking() throws BallerinaTestException {
...
                 new HashMap<>(), projectPath, false);
-        if (!output.contains(msg1) || !output.contains(msg2)) {
+        if (!output.contains(msg1) || !output.contains(msg2) || !output.contains(msg3)) {
             Assert.fail("Test failed due to object mocking failure in test framework.\nOutput:\n" + output);
         }
     }
+
+    @Test()
+    public void testObjectMocking_DefaultIncompatibleArgs() throws BallerinaTestException {
+        String msg = "incorrect type of argument provided at position '1' to mock the function get()";
```
Comment (Dilhasha) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java:77]:
```suggestion
        String msg = "incorrect type of argument provided at position '1' to mock the function 'get()'";
```

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java [hunk 6]
```diff
@@ -61,11 +61,93 @@ public void testFunctionMocking() throws BallerinaTestException {
...
+        String[] args = mergeCoverageArgs(new String[]{"--tests", "object_mocking:testDefaultIncompatibleArgs"});
+        String output =
+                balClient.runMainAndReadStdOut("test", args, new HashMap<>(),
+                        projectBasedTestsPath.resolve("object-mocking-tests").toString(), false);
+        if (!output.contains(msg)) {
+            throw new BallerinaTestException(
+                    "Test failed due to default module single test failure.\nOutput:\n" + output);
```
Comment (Dilhasha) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java:84]: Let's improve the error message to state what is failing. ATM, I'm noticing the same error message in multiple tests.

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java (hunk 5 above)
Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java:77]: To have this change, Ill have to modify the message in the `ObjectMock` class 

https://github.com/ballerina-platform/ballerina-lang/blob/master/misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/ObjectMock.java#L180

Ill proceed with applying this wherever applicable

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/FunctionMock.java (hunk 1 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/mock/FunctionMock.java]: Removed escaped single quotes

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java (hunk 3 above)
Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java]: Removed New line and unnecessary `msg3`

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java (hunk 4 above)
Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java]: Condensed all the tests into one using data providers as suggested

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java (hunk 5 above)
Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java]: Applied single quotes where applicable

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java (hunk 6 above)
Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/MockTest.java]: The test cases are all condensed to one using data providers. I checked what the output for a test failure looks like and it shows which Ballerina test has failed so its not too difficult to identify. The failed test is specified in the output

**Review by Dilhasha — APPROVED** [review]

**Review by pramodya1994 — APPROVED** [review]
LGTM

**Pull request merged**